controllers/apps/transformer_component_prometheus_integration.go

```diff
@@ -0,0 +1,152 @@
+		transCtx.V(1).Info(
+			"Component is in compact mode, no need to create monitor services related objects",
+			"component", client.ObjectKeyFromObject(transCtx.ComponentOrig))
+		return nil
+	}
+
+	synthesizeComp := transCtx.SynthesizeComponent
+	graphCli, _ := transCtx.Client.(model.GraphClient)
+
+	if synthesizeComp.MonitorIntegration == nil {
+		return nil
+	}
+	if err := i.buildPrometheusMonitorService(transCtx, synthesizeComp.MonitorIntegration, graphCli, dag); err != nil {
+		return err
+	}
+	return i.buildVMMonitorService(transCtx, synthesizeComp.MonitorIntegration, graphCli)
+}
+
+func (i componentPrometheusIntegrationTransformer) buildPrometheusMonitorService(transCtx *componentTransformContext, monitorIntegration *appsv1alpha1.MonitorIntegration, graphCli model.GraphClient, dag *graph.DAG) error {
+	var running *monitoringv1.ServiceMonitor
+
+	objects, err := listMonitorServices(transCtx.GetContext(),
+		i.Client,
+		transCtx.Cluster.Name,
+		transCtx.Component.Name,
+		transCtx.Component,
+		intctrlutil.MonitorServiceSignature)
+	if err != nil {
+		return err
+	}
+	if monitorIntegration.ServiceMonitorTemplate == nil {
+		deleteObjects(objects, graphCli, dag)
+		return nil
+	}
+
+	cmp := func(obj monitoringv1.ServiceMonitor) bool {
+		return obj.Namespace == monitorIntegration.ServiceMonitorTemplate.Namespace &&
+			strings.HasPrefix(obj.Name, monitorIntegration.ServiceMonitorTemplate.Name)
+	}
+	index := slices.IndexFunc(objects, cmp)
+	if index >= 0 {
+		running = objects[index].DeepCopy()
+		objects = slices.Delete(objects, index, index+1)
+	}
+
+	deleteObjects(objects, graphCli, dag)
+	return createOrUpdateMonitorService(transCtx, running, monitorIntegration.ServiceMonitorTemplate, graphCli, dag)
+}
+
+func createOrUpdateMonitorService(transCtx *componentTransformContext, existing *monitoringv1.ServiceMonitor, template *appsv1alpha1.ServiceMonitorTemplate, graphCli model.GraphClient, dag *graph.DAG) error {
+	expected, err := createMonitorService(transCtx, template, transCtx.Component)
+	if err != nil {
+		return err
+	}
+
+	if existing == nil {
+		graphCli.Create(dag, expected, inDataContext4G())
+	}
+
+	objCopy := existing.DeepCopy()
+	objCopy.Spec = expected.Spec
+
+	if !reflect.DeepEqual(existing, objCopy) {
+		graphCli.Update(dag, existing, objCopy, inDataContext4G())
+	}
+	return nil
+}
+
+func createMonitorService(transCtx *componentTransformContext, template *appsv1alpha1.ServiceMonitorTemplate, owner client.Object) (*monitoringv1.ServiceMonitor, error) {
+	genName := common.SimpleNameGenerator.GenerateName(template.Name)
+	service := &monitoringv1.ServiceMonitor{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      genName,
+			Namespace: template.Namespace,
+			Labels: map[string]string{
+				constant.AppInstanceLabelKey:       transCtx.Cluster.Name,
+				constant.KBAppShardingNameLabelKey: transCtx.Component.Name,
+			},
+		},
+		Spec: template.ServiceMonitorSpec,
```

Comment:
shall we set serviceMonitorSpec.endpoints if user doesn't fill it.

Reply:
done